File: assign1/src/matrixproduct.java
import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;

public class matrixproduct {
    public static void main(String[] args) {
        // Run tests for different matrix sizes
        runTests1();
        runTests2();
        // runTests3();
    }

    public static void runTests1() {
        System.out.println("Running tests for function 1\n");

        try {
            FileWriter fileWriter = new FileWriter("data/data_java1_test.txt");
            BufferedWriter bufferedWriter = new BufferedWriter(fileWriter);

            for (int size = 600; size <= 1000; size += 400) {
                for (int attempt = 1; attempt <= 4; attempt++) {
                    performMultiplication(size, size, bufferedWriter);
                }
            }

            bufferedWriter.close();
        } catch (IOException e) {
            System.out.println("Error writing to file: " + e.getMessage());
        }

    }

    public static void runTests2() {
        System.out.println("Running tests for function 2\n");

        try {
            FileWriter fileWriter = new FileWriter("data/data_java2_test.txt");
            BufferedWriter bufferedWriter = new BufferedWriter(fileWriter);

            for (int size = 600; size <= 3000; size += 400) {
                for (int attempt = 1; attempt <= 4; attempt++) {
                    performLineMultiplication(size, size, bufferedWriter);
                }
            }

            bufferedWriter.close();
        } catch (IOException e) {
            System.out.println("Error writing to file: " + e.getMessage());
        }
    }

    // public static void runTests3() {
    //     System.out.println("Running tests for function 3\n");

    //     try {
    //         FileWriter fileWriter = new FileWriter("data/data_java3_test.txt");
    //         BufferedWriter bufferedWriter = new BufferedWriter(fileWriter);

    //         for (int size = 4096; size <= 10240; size += 2048) {
    //             int[] array = {128, 256, 516};
    //             for (int blockSize : array) {
    //                 for (int attempt = 1; attempt <= 4; attempt++) {
    //                     performBlockMultiplication(size, size, blockSize, bufferedWriter);
    //                 }
    //             }
    //         }

    //         bufferedWriter.close();
    //     } catch (IOException e) {
    //         System.out.println("Error writing to file: " + e.getMessage());
    //     }
    // }

    public static void performMultiplication(int m_ar, int m_br, BufferedWriter bufferedWriter) {

        double[][] pha = new double[m_ar][m_ar];
        double[][] phb = new double[m_ar][m_ar];
        double[][] phc = new double[m_ar][m_ar];

        for (int i = 0; i < m_ar; i++)
            for (int j = 0; j < m_ar; j++)
                pha[i][j] = 1.0;

        for (int i = 0; i < m_br; i++)
            for (int j = 0; j < m_br; j++)
                phb[i][j] = i + 1;

        long startTime = System.currentTimeMillis();

        for (int i = 0; i < m_ar; i++) {
            for (int j = 0; j < m_br; j++) {
                double temp = 0;
                for (int k = 0; k < m_ar; k++) {
                    temp += pha[i][k] * phb[k][j];
                }
                phc[i][j] = temp;
            }
        }

        long endTime = System.currentTimeMillis();
        System.out.println("Time: " + (endTime - startTime) / 1000.0 + " seconds");

        // Write results to the file
        try {
            bufferedWriter.write("Dimensions: " + m_ar);
            bufferedWriter.newLine();
            bufferedWriter.write("Time: " + (endTime - startTime) / 1000.0 + " seconds");
            bufferedWriter.newLine();
            bufferedWriter.newLine();
        } catch (IOException e) {
            System.out.println("Error writing to file: " + e.getMessage());
        }
    }


    public static void performLineMultiplication(int m_ar, int m_br, BufferedWriter bufferedWriter) {

        double[][] pha = new double[m_ar][m_ar];
        double[][] phb = new double[m_ar][m_ar];
        double[][] phc = new double[m_ar][m_ar];

        for (int i = 0; i < m_ar; i++)
            for (int j = 0; j < m_ar; j++)
                pha[i][j] = 1.0;

        for (int i = 0; i < m_br; i++)
            for (int j = 0; j < m_br; j++)
                phb[i][j] = i + 1;

        long startTime = System.currentTimeMillis();

        for (int i = 0; i < m_ar; i++) {
            for (int k = 0; k < m_ar; k++) {
                for (int j = 0; j < m_br; j++) {
                    phc[i][j] += pha[i][k] * phb[k][j];
                }
            }
        }

        long endTime = System.currentTimeMillis();
        System.out.println("Time: " + (endTime - startTime) / 1000.0 + " seconds");

        // Write results to the file
        try {
            bufferedWriter.write("Dimensions: " + m_ar);
            bufferedWriter.newLine();
            bufferedWriter.write("Time: " + (endTime - startTime) / 1000.0 + " seconds");
            bufferedWriter.newLine();
            bufferedWriter.newLine();
        } catch (IOException e) {
            System.out.println("Error writing to file: " + e.getMessage());
        }

        System.out.println("Result Matrix: ");
        for (int c = 0; c < Math.min(10, m_br); c++) {
            System.out.print((int)phc[0][c] + " ");
        }
        System.out.println();

    }


    public static void performBlockMultiplication(int m_ar, int m_br, int blockSize, BufferedWriter bufferedWriter) {

        double[][] pha = new double[m_ar][m_ar];
        double[][] phb = new double[m_ar][m_ar];
        double[][] phc = new double[m_ar][m_ar];

        for (int i = 0; i < m_ar; i++)
            for (int j = 0; j < m_ar; j++)
                pha[i][j] = 1.0;

        for (int i = 0; i < m_br; i++)
            for (int j = 0; j < m_br; j++)
                phb[i][j] = i + 1;

        long startTime = System.currentTimeMillis();

        for (int a = 0; a < m_ar; a += blockSize) {
            for (int c = 0; c < m_ar; c += blockSize) {
                for (int b = 0; b < m_br; b += blockSize) {
                    for (int i = a; i < a + blockSize; i++) {
                        for (int k = c; k < c + blockSize; k++) {
                            for (int j = b; j < b + blockSize; j++) {
                                phc[i][j] += pha[i][k] * phb[k][j];
                            }
                        }
                    }
                }
            }
        }

        long endTime = System.currentTimeMillis();
        System.out.println("Time: " + (endTime - startTime) / 1000.0 + " seconds");

        // Write results to the file
        try {
            bufferedWriter.write("Block size: " + blockSize);
            bufferedWriter.newLine();
            bufferedWriter.write("Dimensions: " + m_ar);
            bufferedWriter.newLine();
            bufferedWriter.write("Time: " + (endTime - startTime) / 1000.0 + " seconds");
            bufferedWriter.newLine();
            bufferedWriter.newLine();
        } catch (IOException e) {
            System.out.println("Error writing to file: " + e.getMessage());
        }

        System.out.println("Result Matrix: ");
        for (int c = 0; c < Math.min(10, m_br); c++) {
            System.out.print((int)phc[0][c] + " ");
        }
        System.out.println();
    }
}

File: assign1/src/matrixproduct.cpp
#include <stdio.h>
#include <iostream>
#include <iomanip>
#include <time.h>
#include <cstdlib>
#include <fstream>
#include <string>
#include <papi.h>
#include <omp.h>

using namespace std;
#define SYSTEMTIME clock_t

void OnMult(int m_ar, int m_br, ofstream& outputFile) 
{
	SYSTEMTIME Time1, Time2;
	
	char st[100];
	double temp;
	int i, j, k;

	double *pha, *phb, *phc;
		
    pha = (double *)malloc((m_ar * m_ar) * sizeof(double));
	phb = (double *)malloc((m_ar * m_ar) * sizeof(double));
	phc = (double *)malloc((m_ar * m_ar) * sizeof(double));

	for(i=0; i<m_ar; i++)
		for(j=0; j<m_ar; j++)
			pha[i*m_ar + j] = (double)1.0;

	for(i=0; i<m_br; i++)
		for(j=0; j<m_br; j++)
			phb[i*m_br + j] = (double)(i+1);

    Time1 = clock();

	for(i=0; i<m_ar; i++) {
        for( j=0; j<m_br; j++){
            temp = 0;
            for( k=0; k<m_ar; k++) {	
                temp += pha[i*m_ar+k] * phb[k*m_br+j];
            }
            phc[i*m_ar+j]=temp;
        }
	}

    Time2 = clock();
	sprintf(st, "Time: %3.3f seconds\n", (double)(Time2 - Time1) / CLOCKS_PER_SEC);
	outputFile << st;
    cout << st;
	

	// display 10 elements of the result matrix tto verify correctness
	cout << "Result matrix: " << endl;
	for(i=0; i<1; i++)
	{	for(j=0; j<min(10,m_br); j++)
			cout << phc[j] << " ";
	}
	cout << endl;

    free(pha);
    free(phb);
    free(phc);
}


void OnMultLine(int m_ar, int m_br, ofstream& outputFile)
{
    SYSTEMTIME Time1, Time2;
    char st[100];
    int i, j, k;

    double *pha, *phb, *phc;
        
    pha = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phb = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phc = (double *)malloc((m_ar * m_ar) * sizeof(double));

    for(i=0; i<m_ar; i++)
        for(j=0; j<m_ar; j++)
            pha[i*m_ar + j] = (double)1.0;

    for(i=0; i<m_br; i++)
        for(j=0; j<m_br; j++)
            phb[i*m_br + j] = (double)(i+1);
    
    for(i=0; i<m_br; i++)
        for(j=0; j<m_br; j++)
            phc[i*m_br + j] = (double)0.0;

    Time1 = clock();

    for(i=0; i<m_ar; i++) {
        for( k=0; k<m_ar; k++ ) {
            for( j=0; j<m_br; j++) {    
                phc[i*m_ar+j] += pha[i*m_ar+k] * phb[k*m_br+j];
            }
        }
    }

    Time2 = clock();
    sprintf(st, "Time: %3.3f seconds\n", (double)(Time2 - Time1) / CLOCKS_PER_SEC);
    outputFile << st;
    cout << st;
    
    // display 10 elements of the result matrix to verify correctness
    cout << "Result matrix: " << endl;
    for(i=0; i<1; i++)
    {   for(j=0; j<min(10,m_br); j++)
			cout << phc[j] << " ";
    }
    cout << endl;

    free(pha);
    free(phb);
    free(phc);
}


void OnMultBlock(int m_ar, int m_br, int bkSize, ofstream& outputFile)
{
    SYSTEMTIME Time1, Time2;
    char st[100];
    int a, b, c, i, j, k;

    double *pha, *phb, *phc;
        
    pha = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phb = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phc = (double *)malloc((m_ar * m_ar) * sizeof(double));

    for(i=0; i<m_ar; i++)
        for(j=0; j<m_ar; j++)
            pha[i*m_ar + j] = (double)1.0;

    for(i=0; i<m_br; i++)
        for(j=0; j<m_br; j++)
            phb[i*m_br + j] = (double)(i+1);
    
    for(i=0; i<m_br; i++)
        for(j=0; j<m_br; j++)
            phc[i*m_br + j] = (double)0.0;

    Time1 = clock();

    for (a = 0; a < m_ar; a += bkSize) {
        for (c = 0; c < m_ar; c += bkSize) {
            for (b = 0; b < m_br; b += bkSize) {
                for(i = a; i < a + bkSize; i++) {
                    for(k = c; k < c + bkSize; k++ ) {
                        for(j = b; j < b + bkSize; j++) {    
                            phc[i*m_ar+j] += pha[i*m_ar+k] * phb[k*m_br+j];
                        }
                    }
                }
            }
        }
    }

    Time2 = clock();
    sprintf(st, "Time: %3.3f seconds\n", (double)(Time2 - Time1) / CLOCKS_PER_SEC);
    outputFile << st;
    cout << st;
    
    // display 10 elements of the result matrix to verify correctness
    cout << "Result matrix: " << endl;
    for(i=0; i<1; i++)
    {   for(j=0; j<min(10,m_br); j++)
			cout << phc[j] << " ";
    }
    cout << endl;

    free(pha);
    free(phb);
    free(phc);
}


void OnMultLineParallel1(int m_ar, int m_br, ofstream& outputFile)
{
    double Time1, Time2;
    char st[100];
    int i, j;

    double *pha, *phb, *phc;
        
    pha = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phb = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phc = (double *)malloc((m_ar * m_ar) * sizeof(double));

    for(i=0; i<m_ar; i++)
        for(j=0; j<m_ar; j++)
            pha[i*m_ar + j] = (double)1.0;

    for(i=0; i<m_br; i++)
        for(j=0; j<m_br; j++)
            phb[i*m_br + j] = (double)(i+1);
    
    for(i=0; i<m_br; i++)
        for(j=0; j<m_br; j++)
            phc[i*m_br + j] = (double)0.0;

    Time1 = omp_get_wtime();

    #pragma omp parallel for
    for(int i=0; i<m_ar; i++) {
        for(int  k=0; k<m_ar; k++ ) {
            for(int j=0; j<m_br; j++) {    
                phc[i*m_ar+j] += pha[i*m_ar+k] * phb[k*m_br+j];
            }
        }
    }

    Time2 = omp_get_wtime();
    sprintf(st, "Time: %3.3f seconds\n", Time2 - Time1);
    outputFile << st;
    cout << st;
    
    // display 10 elements of the result matrix to verify correctness
    cout << "Result matrix: " << endl;
    for(i=0; i<1; i++)
    {   for(j=0; j<min(10,m_br); j++)
			cout << phc[j] << " ";
    }
    cout << endl;

    free(pha);
    free(phb);
    free(phc);
}


void OnMultLineParallel2(int m_ar, int m_br, ofstream& outputFile)
{
    double Time1, Time2;
    char st[100];
    int i, j;

    double *pha, *phb, *phc;
        
    pha = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phb = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phc = (double *)malloc((m_ar * m_ar) * sizeof(double));

    for(i=0; i<m_ar; i++)
        for(j=0; j<m_ar; j++)
            pha[i*m_ar + j] = (double)1.0;

    for(i=0; i<m_br; i++)
        for(j=0; j<m_br; j++)
            phb[i*m_br + j] = (double)(i+1);
    
    for(i=0; i<m_br; i++)
        for(j=0; j<m_br; j++)
            phc[i*m_br + j] = (double)0.0;

    Time1 = omp_get_wtime();

    #pragma omp parallel
    for(int i=0; i<m_ar; i++) {
        for(int k=0; k<m_ar; k++ ) {
            #pragma omp for
            for(int j=0; j<m_br; j++) {    
                phc[i*m_ar+j] += pha[i*m_ar+k] * phb[k*m_br+j];
            }
        }
    }

    Time2 = omp_get_wtime();
    sprintf(st, "Time: %3.3f seconds\n", Time2 - Time1);
    outputFile << st;
    cout << st;
    
    // display 10 elements of the result matrix to verify correctness
    cout << "Result matrix: " << endl;
    for(i=0; i<1; i++)
    {   for(j=0; j<min(10,m_br); j++)
			cout << phc[j] << " ";
    }
    cout << endl;

    free(pha);
    free(phb);
    free(phc);
}


void handle_error(int retval)
{
    printf("PAPI error %d: %s\n", retval, PAPI_strerror(retval));
    exit(1);
}

int main(int argc, char *argv[])
{   
    int op = atoi(argv[1]);

    ofstream f;
    f.open(op == 3 ? argv[4] : argv[3], ios::app);

    int matrix_size = atoi(argv[2]);

	int size = atoi(argv[2]);
    
    // PAPI setup
    int EventSet = PAPI_NULL;
    long long values[3];
    int ret;
    ret = PAPI_library_init(PAPI_VER_CURRENT);
    if (ret != PAPI_VER_CURRENT)
    {
        cout << "FAIL" << endl;
    }

    ret = PAPI_create_eventset(&EventSet);
    if (ret != PAPI_OK)
        cout << "ERROR: create eventset" << endl;

    ret = PAPI_add_event(EventSet, PAPI_L1_DCM);
    if (ret != PAPI_OK)
        cout << "ERROR: PAPI_L1_DCM" << endl;

    ret = PAPI_add_event(EventSet, PAPI_L2_DCM);
    if (ret != PAPI_OK)
        cout << "ERROR: PAPI_L2_DCM" << endl;

    ret = PAPI_start(EventSet);
	if (ret != PAPI_OK)
		cout << "ERROR: Start PAPI" << endl;

    switch (op)
    {
    case 1:
		OnMult(size, size, f);	
        break;
    case 2:
		OnMultLine(size, size, f);            
        break;
    case 3:
    {
        int block_size = atoi(argv[3]);
        cout << "Block size: " << block_size << endl;
        f << "Block size: " << block_size << endl;
        OnMultBlock(size, size, block_size, f);
        break;
    }
    case 4:
        OnMultLineParallel1(size, size, f);
        break;
    case 5:
        OnMultLineParallel2(size, size, f);
        break;
    default:
        cout << "Invalid option" << endl;
        break;
    }

    ret = PAPI_stop(EventSet, values);
	if (ret != PAPI_OK)
	    cout << "ERROR: Stop PAPI" << endl;

	printf("L1 DCM: %lld \n", values[0]);
	printf("L2 DCM: %lld \n", values[1]);


	f << "Dimensions: " << matrix_size << "\n";
	f << "L1 DCM:" <<  values[0] << "\n";
	f << "L2 DCM:" <<  values[1] << "\n\n";

    f.close();

    ret = PAPI_reset(EventSet);
    if (ret != PAPI_OK)
        cout << "FAIL reset" << endl;

    ret = PAPI_remove_event(EventSet, PAPI_L1_DCM);
    if (ret != PAPI_OK)
        cout << "FAIL remove event" << endl;

    ret = PAPI_remove_event(EventSet, PAPI_L2_DCM);
    if (ret != PAPI_OK)
        cout << "FAIL remove event" << endl;

    ret = PAPI_destroy_eventset(&EventSet);
    if (ret != PAPI_OK)
        cout << "FAIL destroy" << endl;

    return 0;
}
